particle-iot/device-os
# Ledger API and sync

### Problem

This PR supersedes https://github.com/particle-iot/device-os/pull/2660 and https://github.com/particle-iot/device-os/pull/2675.

Below are some links to simplify navigation in the new code:

- [Ledger application API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_ledger.h)
- [Ledger system API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/inc/system_ledger.h)
- [Ledger storage layer](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/src/ledger/ledger.cpp)
- [Ledger state machine](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/system/src/ledger/ledger_manager.cpp)
- [Synchronization protocol](https://github.com/particle-iot/device-os-protobuf/blob/main/cloud/ledger.proto)
- [New CoAP API](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/communication/inc/coap_api.h)
- [Variant class](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_variant.h)
- [Map class](https://github.com/particle-iot/device-os/blob/ledger-sync-merged/sc-120056/wiring/inc/spark_wiring_map.h)

### Steps to Test

See the [validation plan](https://www.notion.so/particle/Ledger-Validation-7b64402038734c28af40b2531da8a323).

The integration tests in this PR require the device's product to have the following ledger definitions:

- `test-device-to-cloud` – type: Device to Cloud.
- `test-cloud-to-device` – type: Cloud to Device; scope: Device.

### References

- sc-118798
- sc-119316
- sc-120056
- sc-122990
- sc-124092
